syntax = "proto3";

package envoy.extensions.filters.http.mgw_wasm_websocket.v3;

service WebSocketFrameService {
  rpc PublishFrameData(stream WebSocketFrameRequest) returns (stream WebSocketFrameResponse) {}    
}

message WebSocketFrameRequest {
  string node_id = 1;

  Metadata metadata = 2;
  
  int32 frame_length = 3;

  string remote_ip = 4;

  bytes payload = 5;

  // direction of the message
  enum MessageDirection {
        PUBLISH = 0;
        SUBSCRIBE = 1;
        HANDSHAKE = 2;
  }
  MessageDirection direction = 6;

  int32 apim_error_code = 7;
}

message WebSocketFrameResponse {
  enum Code {
        // The response code is not known.
        UNKNOWN = 0;
        // The response code to notify that the number of requests are under limit.
        OK = 1;
        // The response code to notify that the number of requests are over limit.
        OVER_LIMIT = 2;
    }

  Code throttle_state = 1;

  int64 throttle_period = 2;

  int32 apim_error_code = 3;
}

// message MetadataValue {
//   string key = 1;
//   string value = 2;
// }

message Metadata {
  map<string,string> ext_authz_metadata = 1;
}
